Make IndexShard.resetEngine re-throw on failures

yungene opened the pull request:
Updated IndexShard.resetEngine to re-throw on failures to bubble up critical issues back to the caller.

Relates ES-10705

yungene committed:
Make IndexShard.resetEngine rethrow on failures

Updated IndexShard.resetEngine to rethrow on failures to bubble up
critical issues back to the caller.

yungene committed:
Merge branch 'main' into hollow-it/stress-test-obs

kingherc commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -4667,16 +4671,26 @@ public void resetEngine(Consumer<Engine> postResetNewEngineConsumer) {
                     failShard("unable to close previous engine after reset", e);
+
+                    if (primaryException != null) {
+                        primaryException.addSuppressed(e);
```
nit I had to read to understand what happens with exceptions in finally. If I'm understanding things correctly, any thrown exception in finally will hide any previous one that was thrown. So it seems Java prioritizes the finally exception over the previous one. So I wonder if we'd like to honor Java prioritization, and instead do here:

```
e.addSuppressed(primaryException);
throw e;
```

yungene replied:
I believe the standard way is to throw the original exception and add the extra exceptions from finally as suppressed. This way the root cause would come first. But we can do as you say as make the exception from finally be the top exception.

In practice, to have an exception in finally, we must have `previousEngine != null`, which means that the try block must mostly succeed. So we don't really expect to hit that sequence of events at all. The only annoyance is that Intellij highlights this as a warning. If you think that would make it easier to understand I can do that. But I don't think we should ignore primary exception altogether. We want to avoid missing it with any future changes.

yungene commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -4634,16 +4634,20 @@ public void afterRefresh(boolean didRefresh) {
      *                                   The provided engine reference should not be retained by the consumer.
      */
-    public void resetEngine(Consumer<Engine> postResetNewEngineConsumer) {
+    public void resetEngine(Consumer<Engine> postResetNewEngineConsumer) throws IOException {
```
Are you suggestion to instead declare `throws Exception`? I think that would indeed be clearer.

IOException is an Exception, so the catch works just fine. But `engine.prepareForEngineReset();` and `IOUtils.close(previousEngine);` declare to throw IOException. So we need to either transform it to an unchecked exception or declare to throw it. But also we can declare to throw a more general Exception to be less ambiguous.

kingherc reviewed (commented):
Thanks, just one comment before approving.

yungene committed:
Update resetEngine and relocated signatures

Make IndexShard.resetEngine() throw Exception instead of IOException.

Change IndexShard.relocated() to take CheckedBiConsumer instead of
BiConsumer. This helps using resetEngine() inside of consumer if needed.

fcofdez reviewed (approved):
LGTM

kingherc reviewed (approved):
LGTM, thanks for taking care of this!

yungene committed:
Docs change addressing PR comms

yungene committed:
Merge branch 'main' into hollow-it/stress-test-obs
